cardano-testnet/src/Testnet/Runtime.hs

```diff
@@ -204,14 +224,30 @@ startLedgerNewEpochStateLogging testnetRuntime tmpWorkspace = withFrozenCallStac
           Api.QuickValidation
           (EpochNo maxBound)
           ()
-          (\epochState _ _ -> handler logFile epochState)
+          (\epochState _ _ ->
+              liftIO $ evalStateT (handler logFile epochState) 0
+          )
       H.note_ $ "Started logging epoch states to: " <> logFile
+
+      -- In order to create a diff of the epoch state logging file contents
+      -- we must do so when the resources are deallocated (see runInBackground)
+      -- and therefore when the log file is no longer in use.
+      void . H.evalM $ allocate (pure ()) $ \_ -> do
+          isFileReadableLoop logFile >>= \case
+            False -> error "isFileReadableLoop: Impossible"
+            True -> do
+              logFileContents <- IO.readFile logFile
+              let epochStateValues = epochStateBeforeAfterValues logFileContents
+                  epochStateDiffs' = epochStateDiffs epochStateValues
+              Text.writeFile (logDir </> "ledger-epoch-state-diffs.log") epochStateDiffs'
+
   where
-    handler :: FilePath -> AnyNewEpochState -> StateT () IO LedgerStateCondition
-    handler outputFp anyNewEpochState = handleException . liftIO $ do
-      appendFile outputFp $ "#### BLOCK ####" <> "\n"
-      appendFile outputFp $ show anyNewEpochState <> "\n"
-      pure ConditionNotMet
+    handler :: FilePath -> AnyNewEpochState -> StateT Int IO LedgerStateCondition
+    handler outputFp (AnyNewEpochState sbe nes) = do
+      handleException . liftIO $ do
+        appendFile outputFp $ "#### BLOCK ####" <> "\n"
```

Comment:
Can we make this tag a json too? I'm thinking about the use case, where we could filter the json logs through `jq`, so a non-json line would break jq afair.

Reply:
Yeah, `jsonl` is better than `json`.